### ISSUE TITLE
Block elements by size

### ISSUE THREAD

**Issue by wavexx** [gorhill/uBlock#1163]
I'm just throwing an idea here.. is it actually possible in FF to block (or better to say, stop) downloading elements if the size is known immediately in the Content-Lenght header?

I'm looking at those huge media elements which are now popular in mainstream websites. With my slow network connection, any element in the page beyond 1mb is not worth *attempting* to download. But it doesn't stop at this: I've seen background images talling at 4mb, or poorly compressed fonts, or worse yet movies as background!! It's insane.

So blocking by media type wouldn't be always correct (although I do miss sometimes the umatrix interface, I rely too much on cosmetic fixes to use it).

Although I wonder how such mechanism would play with user-initiated downloads, if at all. In any case, a "toggle" would still be super-useful for my purpose.

**Comment by gwarser** [comment]
I would also like this.
This will require inspecting **response** headers. Will not break _attempts_ to download, but will abort downloading huge resources.
Probably not job for uBO, but if someone will create extension for this, I will use it.

**Comment by gorhill** [comment]
I can see the usefulness. I will think about it.

**Comment by wavexx** [comment]
On 03/01/16 22:09, Raymond Hill wrote:
> I can see the usefulness. I will think about it.

Like gwarser said, I also don't need such thing to be integrated in uB,
could be detached as well.

I was hoping somebody would have some experience with what FF already
provides to know if that can be implemented or not.

Right now I do this using a local proxy, but it's not convenient. If I
need to download something it's cumbersome. I'd rather press a button on
the toolbar to disable the limit temporarily.

**Comment by gorhill** [comment]
Good test case: <http://www.wired.com/2016/01/drones-arent-just-toys-anymore/>. Does not add any value to the article, yet consumes a good chunk of bandwidth, 3,612,224 bytes (tested on desktop).

**Comment by wavexx** [comment]
On 08/01/16 17:54, Raymond Hill wrote:
> Good test case:
> http://www.wired.com/2016/01/drones-arent-just-toys-anymore/. Does not
> add any value to the article, yet consumes a good chunk of bandwidth,
> 3,612,224 bytes (tested on desktop).

If you start looking, you'll find plenty.
Unfortunately, I notice mainly due to my slow connection.

The irony? Sites like google pagespeed generally penalize the rating if
you have images which can be recompressed with lower settings, but do
not actually factor in the sizes (or count) of the elements themselves.

In fact, pagespeed is perfectly happy to give you a score of B/C if you
have less than 2kb of elements which are not minified while *still*
loading in sub-second time on GPRS, while giving A to a site that loads
5mb+ of elements scattered via JS async loading and are fetched from 10
different subdomains, with countless font faces, but are all compressed
and minified with maximum settings.

Riiight.

**Comment by gorhill** [comment]
Will just put in here examples of how this can help:

<http://www.wired.com/2016/01/drones-arent-just-toys-anymore>
- Not blocking large media elements: 5.5 MB.
- Blocking media elements larger than 50 KB: 1.2 MB.
- Blocking media elements larger than 1 MB: 2.1 MB.


***

Test done with dev console opened and cache disabled.

**Comment by gorhill** [comment]
For translators:

![a](https://cloud.githubusercontent.com/assets/585534/12380164/2575ee24-bd3a-11e5-8743-24da038463f8.png)

***

![c](https://cloud.githubusercontent.com/assets/585534/12380165/296a171c-bd3a-11e5-9914-660037e961a5.png)

***

![d](https://cloud.githubusercontent.com/assets/585534/12380166/2bdd2606-bd3a-11e5-89ac-eda22e9c0e36.png)

**Comment by gorhill** [comment]
Fixed with 89148351e86ce933ba5fc8cc998681a884c13a3f.

**Comment by gwarser** [comment]
I checked this now, and I see problem - very often all images are blocked. For example all avatars on github (here or main page). Should I open new issue?

![ublock1](https://cloud.githubusercontent.com/assets/886325/12380372/996fa75c-bd71-11e5-93c1-f32fc1a0dc0c.png)

**Comment by gorhill** [comment]
> Should I open new issue?

No. I have been testing this for two weeks now. Use the dev console (Network tab can shoe response headers) to confirm there is something wrong first, and if you confirm, provide an exact example of the

**Comment by gwarser** [comment]
Firefox 64b 43, 44, 45 on Linux Mint.

Fx 43 refreshed by about:support -> Refresh Firefox and:
Only uBlock0 enabled.
Open this page
Open http://www.wired.com/2016/01/drones-arent-just-toys-anymore in background by middle click -> OK
Close and open again -> images missing

Browser log: maybe CSP or maybe 'unknown network error'.
uBlock log: weird, images allowed, then blocked by 'large-media', then allowed again ???

This log contains data from page reloaded after the fact:
[logs.zip](https://github.com/gorhill/uBlock/files/93688/logs.zip)

**Comment by gorhill** [comment]
Most important information:
- Are large media elements blocked by default?
- What is the threshold?

I set to block by default with a threshold of 100 kB, then middle-clicked the Wired link, and the page opened in a new tab, and there were 10 images blocked -- so it worked as expected. It's not clear what you are reporting. You say _"middle click -> OK"_... What does "OK" mean? It's why I need a step by step repro case, with what is expected/unexpected, and what exactly you did at each step, so that there is nothing left to misinterpretation. Now I am not sure what is the issue you see. Also, do you have any other extensions installed? If so, test only with uBO to rule out other extensions interfering.

**Comment by gorhill** [comment]
Example:

- Quit and launch FF 43
- Open to uBO settings
- Block large media elements by default
- Set the threshold to 100 kB
- Middle-click <http://www.wired.com/2016/01/drones-arent-just-toys-anymore>
- Click the newly opened tab to bring it forth
- Result: 43 large media elements are reported blocked
- Large image at the top of article is blocked

These were my steps.

**Comment by gwarser** [comment]
Blocked by default with threshold 100kB.

OK - means that large images are blocked (10 reported in popup) eg. this one in header. Thumbnails are not blocked (floating on the right and on the bottom)

When opened again, popup reports  24 (other time this was 40) blocked, and i see only one thumbnail floating on the right, and one under play button on bottom (or none at all).

I cannot access http headers in browser log (don't know why), but dev tools works, and on reload I see this for properly blocked image:
![blocked](https://cloud.githubusercontent.com/assets/886325/12381443/4a317f4e-bd8a-11e5-93f2-4e2b4044aad4.png)

and this for these invisible:
![not-blocked-invisible](https://cloud.githubusercontent.com/assets/886325/12381446/4f16e684-bd8a-11e5-8530-807624abe998.png)

I'm sorry I cannot explain better, I should be sleeping for 4 hours now :)

**Comment by gorhill** [comment]
> on reload I see this for (I think, should be) properly blocked image

What matters is the size of the image -- which I cannot see in your screenshot:

![a](https://cloud.githubusercontent.com/assets/585534/12381561/46fad44a-bd5a-11e5-9598-24cf7dba4ad0.png)

**Comment by gwarser** [comment]
This is large image blocked as expected. You can see in response header `Content-Length: "260182"`.

Problem is that image from second screenshot is also hidden. 

![3](https://cloud.githubusercontent.com/assets/886325/12381633/613ca4a8-bd8e-11e5-8acd-38cf8148c99a.png)

**Comment by gorhill** [comment]
Ok I see. If the `Content-Length` header is not present, uBO assumes the worst, i.e. that the resource is very large. This happens in real cases, the `Content-Length` header is not always present. However in the current case, it is not present because the server is telling the browser to use its cached version. So uBO has to disregard responses for which the `Content-Length` field is not expected.

**Comment by gwarser** [comment]
`Content-Length` may also be missing for chunked encoding.

**Comment by gorhill** [comment]
Ok so I guess I should not make uBO assume the worst to be safer, and bail out if the `Content-Lenght` is not present. It's a convenience feature after all, it's fine if it catches the majority of cases.

**Comment by gwarser** [comment]
I agree.

**Comment by gwarser** [comment]
Works fine now, thank you.

**Comment by ruy-benton** [comment]
Great idea,
I 'm waiting for this ... for long time
Thanks

**Comment by ReticentAdjuvant** [comment]
Omg. I actually downloaded the pre-release here because I wanted to use this feature so bad. Thank you for making it! Would it be possible to alter the threshold per source?
I want to be able to do this in "My Rules" or something to this affect.
```
no-large-media: www.google.com true 50kb
no-large-media: www.fakesite.com true 300kb
no-large-media: www.bacon.com true 2mb
```
I'd ask for an interface that identifies and displays elements from largest to smallest but that might be a bit much. I'm not sure. (Maybe for uMatrix?)

**Comment by epicgirl1998** [comment]
plus-oneing the above idea

on nyaa.se, both the ads and the site images are served with the filename `www-N.png` where `N` is an incrementing number presumably to make it harder to block the ads

i noticed that the ad images are all >25KB while none of the site images exceed 5KB

it would be trivial to write a filter for the ads if ublock supported dynamic `no-large-media` rules
